File: python/437_路径总和_III_medium.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: TreeNode, targetSum: int) -> int:
        prefix = defaultdict(int)
        prefix[0] = 1
        def dfs(root, curr):
            if not root:
                return 0
            res = 0
            curr += root.val
            if curr-targetSum in prefix:
                res += prefix[curr-targetSum]
            prefix[curr] += 1
            res += dfs(root.left, curr)
            res += dfs(root.right, curr)
            prefix[curr] -= 1
            return res
        return dfs(root, 0)

File: cpp/437_路径总和_III_medium.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    unordered_map<int, int> prefix;
public:
    int pathSum(TreeNode* root, int targetSum) {
        prefix[0] = 1;
        return dfs(root, 0, targetSum);
    }
    int dfs(TreeNode *root, long long cur, int targetSum) {
        if (!root) return 0;
        int res = 0;
        cur += root->val;
        if (prefix.count(cur-targetSum))
            res = prefix[cur-targetSum];
        prefix[cur]++;
        res += dfs(root->left, cur, targetSum);
        res += dfs(root->right, cur, targetSum);
        prefix[cur]--;
        return res;
    }
};